packages/alchemy/src/client/factory.ts

```diff
@@ -0,0 +1,119 @@
+import {
+  erc4337ClientActions,
+  type UserOperationRequest,
+} from "@alchemy/aa-core";
+import {
+  createClient,
+  http,
+  publicActions,
+  size,
+  slice,
+  type Address,
+  type Chain,
+  type Client,
+  type HttpTransport,
+  type HttpTransportConfig,
+  type Transport,
+} from "viem";
+import type { DebugTransaction, SimulateExecutionResponse } from "./interfaces";
+import type {
+  AlchemyEnhanced4337Client,
+  AlchemyEnhancedApiSchema,
+} from "./types";
+
+export const alchemyEnhancedApiActions = (client: Client) => {
+  const clientAdapter = client as Client<
+    Transport,
+    Chain,
+    undefined,
+    AlchemyEnhancedApiSchema
+  >;
+
+  return {
+    simulateUserOperationExecution(
+      request: UserOperationRequest
+    ): Promise<SimulateExecutionResponse | SimulateExecutionResponse[]> {
+      const { sender, initCode, callData } = request;
+      const entryPointAddress: Address =
+        "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789";
+
+      const innerTransactionObject: DebugTransaction = {
+        from: entryPointAddress,
+        to: sender,
+        data: callData,
+      };
+
+      if (initCode === "0x") {
+        return clientAdapter.request({
+          method: "alchemy_simulateExecution",
+          params: [innerTransactionObject],
+        });
+      }
+
+      /**
+       * @see {https://eips.ethereum.org/EIPS/eip-4337#client-behavior-upon-receiving-a-useroperation}
+       *
+       * "If initCode is not empty, parse its first 20 bytes as a factory address.
+       * Record whether the factory is staked, in case the later simulation indicates that it needs to be.
+       * If the factory accesses global state, it must be staked - see reputation, throttling and banning section for details."
+       */
+      if (size(initCode) !== 42) {
+        throw new Error(
+          `Invalid init code provided. Expected 0x or 42 bytes, got ${initCode}`
+        );
+      }
+
+      const factoryAddress = slice(initCode, 2, 22);
+      const initCalldata = slice(initCode, 22);
```

Comment:
we have a method on base account for this already. we should break out the contents of the base account impl into a util function and use that here

Reply:
https://github.com/alchemyplatform/aa-sdk/blob/e7a39f50a71aa21aa7ae73e3d704c4a6709aeb18/packages/core/src/account/base.ts#L338

should create a common util to be used for both.

Reply:
ah right i forgot about this. will refactor